File: sdk/dart/saturn.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

typedef EventHandler = void Function(String stream, Map<String, dynamic> payload);

bool _matchesPattern(String pattern, String stream) {
  if (pattern == stream) return true;
  if (pattern.endsWith('*')) return stream.startsWith(pattern.substring(0, pattern.length - 1));
  return false;
}

class SaturnDBClient {
  final String _host;
  final int    _port;
  final String _token;

  Socket?  _socket;
  String   _buffer          = '';
  bool     _closing         = false;
  Duration _reconnectDelay  = const Duration(milliseconds: 500);

  final Map<String, EventHandler> _handlers    = {};
  final StreamController<String>  _lineStream  = StreamController.broadcast();
  StreamSubscription?             _subscription;

  SaturnDBClient({String host = '127.0.0.1', int port = 7379, String token = ''})
      : _host  = host,
        _port  = port,
        _token = token;

  Future<void> connect() async {
    _socket = await Socket.connect(_host, _port);
    _socket!.setOption(SocketOption.tcpNoDelay, true);
    _reconnectDelay = const Duration(milliseconds: 500);

    _subscription = _socket!
        .cast<List<int>>()
        .transform(utf8.decoder)
        .listen(_onData, onDone: _onClose, onError: (_) => _onClose());

    if (_token.isNotEmpty) await _send('AUTH $_token');
  }

  Future<void> emit(String stream, Map<String, dynamic> payload) async {
    await _send('EMIT $stream ${jsonEncode(payload)}');
  }

  Future<Map<String, dynamic>?> get(String stream) async {
    final response = await _send('GET $stream');
    if (response.startsWith('VALUE ')) {
      return jsonDecode(response.substring(6)) as Map<String, dynamic>;
    }
    return null;
  }

  Future<List<Map<String, dynamic>>> since(String stream, int ts) async {
    final response = await _send('SINCE $stream $ts');
    if (response == 'EMPTY') return [];
    return response
        .split('\n')
        .where((l) => l.startsWith('EVENT '))
        .map((l) {
          final rest    = l.substring(6);
          final spaceAt = rest.indexOf(' ');
          return {
            'stream':  rest.substring(0, spaceAt),
            'payload': jsonDecode(rest.substring(spaceAt + 1)),
          };
        })
        .toList();
  }

  Future<void> watch(String pattern, EventHandler handler) async {
    _handlers[pattern] = handler;
    await _send('WATCH $pattern');
  }

  Future<String> ping() => _send('PING');

  Future<void> disconnect() async {
    _closing = true;
    await _subscription?.cancel();
    await _socket?.close();
  }

  Future<String> _send(String line) async {
    _socket!.write('$line\n');
    await for (final response in _lineStream.stream) {
      if (response.startsWith('EVENT ')) {
        _dispatch(response);
        continue;
      }
      return response;
    }
    throw StateError('connection closed');
  }

  void _onData(String chunk) {
    _buffer += chunk;
    final lines = _buffer.split('\n');
    _buffer = lines.removeLast();

    for (final line in lines) {
      final trimmed = line.trim();
      if (trimmed.isEmpty) continue;
      _lineStream.add(trimmed);
    }
  }

  void _dispatch(String line) {
    final rest    = line.substring(6);
    final spaceAt = rest.indexOf(' ');
    final stream  = rest.substring(0, spaceAt);
    final payload = jsonDecode(rest.substring(spaceAt + 1)) as Map<String, dynamic>;

    for (final entry in _handlers.entries) {
      if (_matchesPattern(entry.key, stream)) {
        entry.value(stream, payload);
      }
    }
  }

  void _onClose() {
    if (_closing) return;
    _scheduleReconnect();
  }

  Future<void> _scheduleReconnect() async {
    print('[saturn] disconnected — reconnecting in ${_reconnectDelay.inMilliseconds}ms');
    await Future.delayed(_reconnectDelay);
    _reconnectDelay = Duration(
      milliseconds: (_reconnectDelay.inMilliseconds * 2).clamp(0, 10000),
    );
    try {
      await connect();
      for (final pattern in _handlers.keys) {
        await _send('WATCH $pattern');
      }
      print('[saturn] reconnected');
    } catch (_) {
      _scheduleReconnect();
    }
  }
}

File: sdk/js/saturn.js
class SaturnDBClient {
    #host;
    #socket;
    #buffer;
    #handlers;
    #reconnectDelay;
    #closing;
    #pendingResolve;

    #token;

    constructor(host = "127.0.0.1", port = 7379, token = "") {
        this.#host           = { host, port };
        this.#token          = token;
        this.#socket         = null;
        this.#buffer         = "";
        this.#handlers       = new Map();
        this.#reconnectDelay = 500;
        this.#closing        = false;
        this.#pendingResolve = null;
    }

    async connect() {
        return new Promise((resolve, reject) => {
            const net    = require("net");
            this.#socket = net.createConnection(this.#host, () => {
                this.#reconnectDelay = 500;
                resolve();
            });
            this.#socket.setNoDelay(true);
            this.#socket.on("data",  (data) => this.#onData(data));
            this.#socket.on("error", (err)  => reject(err));
            this.#socket.on("close", ()     => this.#onClose());
        });
        if (this.#token) await this.#sendAndRead(`AUTH ${this.#token}\n`);
    }

    async emit(stream, payload) {
        await this.#sendAndRead(`EMIT ${stream} ${JSON.stringify(payload)}\n`);
    }

    async get(stream) {
        const response = await this.#sendAndRead(`GET ${stream}\n`);
        if (response.startsWith("VALUE ")) return JSON.parse(response.slice(6));
        return null;
    }

    async since(stream, ts) {
        const response = await this.#sendAndRead(`SINCE ${stream} ${ts}\n`);
        if (response === "EMPTY") return [];
        return response
            .split("\n")
            .filter(l => l.startsWith("EVENT "))
            .map(l => {
                const rest    = l.slice(6);
                const spaceAt = rest.indexOf(" ");
                return {
                    stream:  rest.slice(0, spaceAt),
                    payload: JSON.parse(rest.slice(spaceAt + 1)),
                };
            });
    }

    async watch(pattern, handler) {
        this.#handlers.set(pattern, handler);
        await this.#sendAndRead(`WATCH ${pattern}\n`);
    }

    async ping() {
        return this.#sendAndRead("PING\n");
    }

    disconnect() {
        this.#closing = true;
        this.#socket?.destroy();
    }

    #onData(data) {
        this.#buffer += data.toString();
        const lines   = this.#buffer.split("\n");
        this.#buffer  = lines.pop();

        for (const line of lines) {
            if (!line.trim()) continue;
            if (line.startsWith("EVENT ")) {
                this.#dispatchEvent(line);
            } else if (this.#pendingResolve) {
                const resolve        = this.#pendingResolve;
                this.#pendingResolve = null;
                resolve(line.trim());
            }
        }
    }

    #dispatchEvent(line) {
        const rest    = line.slice(6);
        const spaceAt = rest.indexOf(" ");
        const stream  = rest.slice(0, spaceAt);
        const payload = JSON.parse(rest.slice(spaceAt + 1));

        for (const [pattern, handler] of this.#handlers) {
            if (matchesPattern(pattern, stream)) {
                handler({ stream, payload });
            }
        }
    }

    #sendAndRead(line) {
        return new Promise((resolve) => {
            this.#pendingResolve = resolve;
            this.#socket.write(line);
        });
    }

    async #onClose() {
        if (this.#closing) return;
        console.log(`[saturn] disconnected — reconnecting in ${this.#reconnectDelay}ms`);
        await sleep(this.#reconnectDelay);
        this.#reconnectDelay = Math.min(this.#reconnectDelay * 2, 10000);
        try {
            await this.connect();
            for (const pattern of this.#handlers.keys()) {
                await this.#sendAndRead(`WATCH ${pattern}\n`);
            }
            console.log("[saturn] reconnected");
        } catch {
            this.#onClose();
        }
    }
}

function matchesPattern(pattern, stream) {
    if (pattern === stream) return true;
    if (pattern.endsWith("*")) return stream.startsWith(pattern.slice(0, -1));
    return false;
}

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

module.exports = { SaturnDBClient };
